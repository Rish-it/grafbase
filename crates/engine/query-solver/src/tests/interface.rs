use crate::assert_solving_snapshots;

const SCHEMA: &str = r#"
enum join__Graph {
    A @join__graph(name: "a", url: "http://localhost:4200/simple-interface-object/a")
    B @join__graph(name: "b", url: "http://localhost:4200/simple-interface-object/b")
    C @join__graph(name: "c", url: "http://localhost:4200/simple-interface-object/c")
}

type Admin implements Account
    @join__implements(graph: A, interface: "Account")
    @join__type(graph: A, key: "id")
{
    id: ID!
    isMain: Boolean!
    isActive: Boolean!
    name: String!
}

type Query
    @join__type(graph: A)
    @join__type(graph: B)
    @join__type(graph: C)
{
    users: [NodeWithName!]! @join__field(graph: A)
    anotherUsers: [NodeWithName] @join__field(graph: B)
    accounts: [Account] @join__field(graph: B)
}

type Regular implements Account
    @join__implements(graph: A, interface: "Account")
    @join__type(graph: A, key: "id")
{
    id: ID!
    isMain: Boolean!
    name: String!
    isActive: Boolean!
}

type User implements NodeWithName
    @join__implements(graph: A, interface: "NodeWithName")
    @join__type(graph: A, key: "id")
{
    id: ID!
    name: String
    age: Int
    username: String
}

interface Account
    @join__type(graph: A, key: "id")
    @join__type(graph: B, key: "id", isInterfaceObject: true)
    @join__type(graph: C, key: "id", isInterfaceObject: true)
{
    id: ID!
    name: String! @join__field(graph: B)
    isActive: Boolean! @join__field(graph: C)
}

interface NodeWithName
    @join__type(graph: A, key: "id")
    @join__type(graph: B, key: "id", isInterfaceObject: true)
{
    id: ID!
    name: String @join__field(graph: A)
    username: String @join__field(graph: B)
}
"#;

#[test]
fn interface_field_providing_object_field() {
    // age is coming from subgraph A, but needs User.id for this. `anotherUsers` returns an
    // interface though, so need to retrieve the `NodeWithName.id` as an alternative for `User.id`
    assert_solving_snapshots!(
        "interface_field_providing_object_field",
        SCHEMA,
        r#"
        query {
          anotherUsers {
            ... on User {
              age
            }
          }
        }
        "#
    );
}
