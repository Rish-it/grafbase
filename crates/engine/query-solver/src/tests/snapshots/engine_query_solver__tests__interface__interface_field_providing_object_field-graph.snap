---
source: crates/engine/query-solver/src/tests/interface.rs
expression: "digraph {\n    0 [ label = \"root\" ]\n    1 [ label = \"User.age\" ]\n    2 [ label = \"Query.anotherUsers\" ]\n    3 [ label = \"Root#b\", shape=parallelogram, color=dodgerblue ]\n    4 [ label = \"anotherUsers#b\", shape=box, color=dodgerblue ]\n    5 [ label = \"FedEntity#a\", shape=parallelogram, color=dodgerblue ]\n    6 [ label = \"age#a\", shape=box, color=dodgerblue ]\n    8 [ label = \"*NodeWithName.id\" ]\n    9 [ label = \"id#b\", shape=box, color=dodgerblue ]\n    0 -> 2 [ label = \"\" ]\n    0 -> 3 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    0 -> 3 [ label = \"\", style=dashed,arrowhead=none ]\n    3 -> 4 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    4 -> 2 [ label = \"\", color=violet,arrowhead=none ]\n    2 -> 1 [ label = \"\" ]\n    4 -> 5 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    2 -> 5 [ label = \"\", style=dashed,arrowhead=none ]\n    5 -> 6 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    6 -> 1 [ label = \"\", color=violet,arrowhead=none ]\n    2 -> 8 [ label = \"\" ]\n    4 -> 9 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    9 -> 8 [ label = \"\", color=violet,arrowhead=none ]\n    5 -> 8 [ label = \"\", color=orangered,arrowhead=inv ]\n}\n"
---
digraph {
    0 [ root]
    1 [ User.age]
    2 [ Query.anotherUsers]
    3 [ Root#b]
    4 [ anotherUsers#b]
    5 [ FedEntity#a]
    6 [ age#a]
    8 [ *NodeWithName.id]
    9 [ id#b]
    0 -> 2 [ label = "Field" ]
    0 -> 3 [ label = "CreateChildResolver" ]
    0 -> 3 [ label = "HasChildResolver" ]
    3 -> 4 [ label = "CanProvide" ]
    4 -> 2 [ label = "Provides" ]
    2 -> 1 [ label = "Field" ]
    4 -> 5 [ label = "CreateChildResolver" ]
    2 -> 5 [ label = "HasChildResolver" ]
    5 -> 6 [ label = "CanProvide" ]
    6 -> 1 [ label = "Provides" ]
    2 -> 8 [ label = "Field" ]
    4 -> 9 [ label = "CanProvide" ]
    9 -> 8 [ label = "Provides" ]
    5 -> 8 [ label = "Requires" ]
}
